types/joi/index.d.ts

```diff
@@ -13,1330 +13,2207 @@
 //                 Peter Thorson <https://github.com/zaphoyd>
+//                 Will Garcia <https://github.com/thewillg>
 //                 Simon Schick <https://github.com/SimonSchick>
 //                 Alejandro Fernandez Haro <https://github.com/afharo>
+//                 Silas Rech <https://github.com/lenovouser>
+//                 Anand Chowdhary <https://github.com/AnandChowdhary>
+//                 Miro Yovchev <https://github.com/myovchev>
+//                 David Recuenco <https://github.com/RecuencoJones>
+//                 Frederic Reisenhauer <https://github.com/freisenhauer>
+//                 Stefan-Gabriel Muscalu <https://github.com/legraphista>
+//                 Simcha Wood <https://github.com/SimchaWood>
+//                 Steven Barnett <https://github.com/stevendesu>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.4
+// TypeScript Version: 2.8
 
 // TODO express type of Schema in a type-parameter (.default, .valid, .example etc)
 
-export type Types = 'any' | 'alternatives' | 'array' | 'boolean' | 'binary' | 'date' | 'function' | 'lazy' | 'number' | 'object' | 'string';
-
-export type LanguageOptions = string | boolean | null | {
-    [key: string]: LanguageOptions;
-};
-
-export type LanguageRootOptions = {
-    root?: string;
-    key?: string;
-    messages?: { wrapArrays?: boolean; };
-} & Partial<Record<Types, LanguageOptions>> & { [key: string]: LanguageOptions; };
-
-export interface ValidationOptions {
-    /**
-     * when true, stops validation on the first error, otherwise returns all the errors found. Defaults to true.
-     */
-    abortEarly?: boolean;
-    /**
-     * when true, attempts to cast values to the required types (e.g. a string to a number). Defaults to true.
-     */
-    convert?: boolean;
-    /**
-     * when true, allows object to contain unknown keys which are ignored. Defaults to false.
-     */
-    allowUnknown?: boolean;
-    /**
-     * when true, ignores unknown keys with a function value. Defaults to false.
-     */
-    skipFunctions?: boolean;
-    /**
-     * remove unknown elements from objects and arrays. Defaults to false
-     * - when true, all unknown elements will be removed
-     * - when an object:
-     *      - objects - set to true to remove unknown keys from objects
-     */
-    stripUnknown?: boolean | { arrays?: boolean; objects?: boolean };
-    /**
-     * overrides individual error messages. Defaults to no override ({}).
-     */
-    language?: LanguageRootOptions;
-    /**
-     * sets the default presence requirements. Supported modes: 'optional', 'required', and 'forbidden'. Defaults to 'optional'.
-     */
-    presence?: 'optional' | 'required' | 'forbidden';
-    /**
-     * provides an external data set to be used in references
-     */
-    context?: Context;
-    /**
-     * when true, do not apply default values. Defaults to false.
-     */
-    noDefaults?: boolean;
+declare namespace Joi {
```

Comment:
There is no reason to add this extra namespace.

Reply:
It looks like it serves to merge with the value named `Joi` on the second-to-last line.

Reply:
> Shouldn’t this PR also delete `types/hapi__joi`?

I think we should keep it there because it's style available on npm [@hapi/joi](https://www.npmjs.com/package/@hapi/joi), it's just deprecated.